TypeSystem/Reflection/test.cc

```diff
@@ -141,26 +235,100 @@ TEST(Reflection, StructAndVariant) {
+}
+
+TEST(Reflection, SelfContainingStruct) {
+  using namespace reflection_test;
+  using current::reflection::CurrentTypeID;
+  using current::reflection::ReflectedType_Struct;
+  using current::reflection::ReflectedType_Variant;
+
+  const auto typeid_a = static_cast<uint64_t>(CurrentTypeID<SelfContainingA>());
+  const auto typeid_va = static_cast<uint64_t>(CurrentTypeID<std::vector<SelfContainingA>>());
+  const auto typeid_b = static_cast<uint64_t>(CurrentTypeID<SelfContainingB>());
+  const auto typeid_vb = static_cast<uint64_t>(CurrentTypeID<std::vector<SelfContainingB>>());
+  const auto typeid_c = static_cast<uint64_t>(CurrentTypeID<SelfContainingC>());
+  const auto typeid_msc = static_cast<uint64_t>(CurrentTypeID<std::map<std::string, SelfContainingC>>());
+
+  EXPECT_EQ(9203459442620113853ull, typeid_a);
+  EXPECT_EQ(9316106052119012411ull, typeid_va);
+  EXPECT_EQ(9205833655400624647ull, typeid_b);
+  EXPECT_EQ(9316451506340571627ull, typeid_vb);
+  EXPECT_EQ(9208493534322203273ull, typeid_c);
+  EXPECT_EQ(9346663043140503218ull, typeid_msc);
+
   const ReflectedType_Struct& self_a = Value<ReflectedType_Struct>(Reflector().ReflectType<SelfContainingA>());
   EXPECT_EQ(1u, self_a.fields.size());
-  EXPECT_EQ(9206664846159389537ull, static_cast<uint64_t>(self_a.type_id));
-  EXPECT_EQ(9318143269698080259ull, static_cast<uint64_t>(self_a.fields[0].type_id));
+  EXPECT_EQ(typeid_a, static_cast<uint64_t>(self_a.type_id));
+  EXPECT_EQ(typeid_va, static_cast<uint64_t>(self_a.fields[0].type_id));
+
   const ReflectedType_Struct& self_b = Value<ReflectedType_Struct>(Reflector().ReflectType<SelfContainingB>());
   EXPECT_EQ(1u, self_b.fields.size());
-  EXPECT_EQ(9205249121542238939ull, static_cast<uint64_t>(self_b.type_id));
-  EXPECT_EQ(9318143287813964755ull, static_cast<uint64_t>(self_b.fields[0].type_id));
+  EXPECT_EQ(typeid_b, static_cast<uint64_t>(self_b.type_id));
+  EXPECT_EQ(typeid_vb, static_cast<uint64_t>(self_b.fields[0].type_id));
+
   const ReflectedType_Struct& self_c = Value<ReflectedType_Struct>(Reflector().ReflectType<SelfContainingC>());
   EXPECT_EQ(2u, self_c.fields.size());
-  EXPECT_EQ(9200251873128019120ull, static_cast<uint64_t>(self_c.type_id));
-  EXPECT_EQ(9318143287813964755ull, static_cast<uint64_t>(self_c.fields[0].type_id));
-  EXPECT_EQ(9345111461746810545ull, static_cast<uint64_t>(self_c.fields[1].type_id));
+  EXPECT_EQ(typeid_c, static_cast<uint64_t>(self_c.type_id));
+  EXPECT_EQ(typeid_vb, static_cast<uint64_t>(self_c.fields[0].type_id));
+  EXPECT_EQ(typeid_msc, static_cast<uint64_t>(self_c.fields[1].type_id));
+}
 
-  EXPECT_NE(
-      static_cast<uint64_t>(Value<ReflectedType_Variant>(Reflector().ReflectType<FooBarBaz>()).type_id),
-      static_cast<uint64_t>(Value<ReflectedType_Variant>(Reflector().ReflectType<Variant<Foo, Bar, Baz>>()).type_id));
+#ifdef TODO_DKOROLEV_EXTRA_PARANOID_DEBUG_SYMBOL_NAME
```

Comment:
There's some `*_PARANOID_*` symbol in `alt-master` already IIRC, hence I suggest we merge first, and then rename this.